.github/workflows/frontend.yml

```diff
@@ -0,0 +1,222 @@
+            *.cache-to=type=gha,scope=${{ github.workflow }}-${{ env.CACHE_GHA_SCOPE_IT }}
+
+  test:
+    runs-on: ubuntu-20.04
+    needs:
+      - prepare-test
+    strategy:
+      fail-fast: false
+      matrix:
+        pkg:
+          - ./frontend/dockerfile
+        worker:
+          - containerd
+          - containerd-rootless
+          - containerd-1.5
+          - containerd-1.4
+          - containerd-snapshotter-stargz
+          - oci
+          - oci-rootless
+          - oci-snapshotter-stargz
+        typ:
+          - integration
+          - dockerfile
+    steps:
+      -
+        name: Checkout
+        uses: actions/checkout@v3
+      -
+        name: Expose GitHub Runtime
+        uses: crazy-max/ghaction-github-runtime@v2
+      -
+        name: Set up QEMU
+        uses: docker/setup-qemu-action@v2
+      -
+        name: Set up Docker Buildx
+        uses: docker/setup-buildx-action@v2
+        with:
+          version: ${{ env.BUILDX_VERSION }}
+          driver-opts: image=${{ env.REPO_SLUG_ORIGIN }}
+          buildkitd-flags: --debug
+      -
+        name: Test pkg=${{ matrix.pkg }} ; typ=${{ matrix.typ }} ; worker=${{ matrix.worker }}
+        run: |
+          if [ -n "${{ matrix.worker }}" ]; then
+            export TESTFLAGS="${TESTFLAGS} --run=//worker=${{ matrix.worker }}$"
+          fi
+          ./hack/test ${{ matrix.typ }}
+          mv ./coverage/coverage.txt ./coverage/coverage-${{ github.job }}-$(echo "${{ matrix.pkg }}-${{ matrix.typ }}-${{ matrix.worker }}" | tr -dc '[:alnum:]-\n\r' | tr '[:upper:]' '[:lower:]').txt
+        env:
+          TEST_COVERAGE: 1
+          TESTPKGS: ${{ matrix.pkg }}
+          CACHE_FROM: type=gha,scope=${{ env.CACHE_GHA_SCOPE_IT }}
+      -
+        name: Upload coverage file
+        uses: actions/upload-artifact@v3
+        with:
+          name: coverage
+          path: ./coverage
+
+  upload-coverage:
+    runs-on: ubuntu-20.04
+    needs:
+      - test
+    steps:
+      -
+        name: Checkout
+        uses: actions/checkout@v3
+      -
+        name: Download coverage files
+        uses: actions/download-artifact@v3
+        with:
+          name: coverage
+          path: ./coverage
+      -
+        name: List coverage files
+        uses: actions/github-script@v6
+        id: files
+        with:
+          result-encoding: string
+          script: |
+            return require('fs').readdirSync('./coverage', {withFileTypes: true})
+              .filter(item => !item.isDirectory())
+              .map(item => `./coverage/${item.name}`)
+              .join(',');
+      -
+        name: Send to Codecov
+        uses: codecov/codecov-action@v2
+        with:
+          files: ${{ steps.files.outputs.result }}
+
+  image:
+    runs-on: ubuntu-20.04
+    needs:
+      - test
+    strategy:
+      fail-fast: false
+      matrix:
+        channel:
+          - mainline
+          - labs
```

Comment:
There is a slight difference now when we want to release the Dockerfile frontend. Previously we had to push a tag for each channel like `dockerfile/1.4.0`, `dockerfile/1.4.0-labs`. Now you only have to push a single tag `dockerfile/1.4.0` and this will release channels specified in this matrix.